js/src/volume.js

```diff
@@ -485,32 +484,70 @@ var ScatterView = widgets.WidgetView.extend( {
         // colors
         var colors = new THREE.InstancedBufferAttribute(new Float32Array( max_count * 3 ), 3, 1);
         var colors_previous = new THREE.InstancedBufferAttribute(new Float32Array( max_count * 3 ), 3, 1);
-
         function to_rgb(color) {
             color = new THREE.Color(color)
             return [color.r, color.g, color.b]
         }
-        var color = to_rgb(this.model.get("color"))
-        var color_previous = "color" in this.previous_values ? to_rgb(this.previous_values["color"]) : color;
+
+        function get_value_index_color(variable,index,max_count){
+            //Return a two D array (max_count,3)
+         var tmp_color = new Array(max_count);
+
+         if ( typeof variable == "string") {
+           //OD
+            color =  to_rgb(variable)
+            for(var i = 0; i < max_count; i++) {
+              tmp_color[i] = color
+            }
+          }
+
+          else {
+            if ( typeof variable[0] == "string") {
+              // 1 D
+              to_convert = variable
+            }
+            else if (typeof index != "undefined"  && typeof variable[0][0] == "string"){
+              //Two D
+              index1 = Math.min(index,variable.length - 1);
+              to_convert = variable[index1]
+            }
+
+            for(var i = 0; i < max_count; i++) {
+              tmp_color[i] = to_rgb(to_convert[i])
+            }
+          }
+
+          return tmp_color
+        }
+
+
+        var color = get_value_index_color(this.model.get("color"),index,max_count)
+
+        var color_previous = "color" in this.previous_values ? get_value_index_color(this.previous_values["color"],this.previous_values["index"],max_count) : color;
         if(!color_previous)
             color_previous = color;
+        console.log("color",color)
+        console.log("color_previous",color_previous)
 
         var color_selected = to_rgb(this.model.get("color_selected"))
         var color_selected_previous = "color_selected" in this.previous_values ? to_rgb(this.previous_values["color_selected"]) : color_selected;
         if(!color_selected_previous)
             color_selected_previous = color_selected;
 
-	    for(var i = 0; i < max_count; i++) {
-	        var cur_color = color;
-	        if(selected.indexOf(i) != -1)
-	            cur_color = color_selected
-   	        colors.setXYZ(i, cur_color[0], cur_color[1], cur_color[2]);
+  	    for(var i = 0; i < max_count; i++) {
+  	        var cur_color = color[i];
+
+  	        if(selected.indexOf(i) != -1)
+  	            cur_color = color_selected
+
+     	        colors.setXYZ(i, cur_color[0], cur_color[1], cur_color[2]);
+
+  	        var cur_color_previous = color_previous[i];
+  	        if(selected_previous.indexOf(i) != -1)
+  	            cur_color_previous = color_selected_previous
+     	        colors_previous.setXYZ(i, cur_color_previous[0], cur_color_previous[1], cur_color_previous[2]);
+  	    }
 
```

Comment:
This will decrease the performance for having a million points, but a single color. So I'd rather see that if it is a single color, to not create an array.